src/validation.cpp

```diff
@@ -5082,3 +5083,48 @@ void ChainstateManager::MaybeRebalanceCaches()
         }
     }
 }
+
+constexpr int SNAPSHOT_NAME_LEN = 75; // "chainstate_" + 64 hex characters for blockhash.
+
+static bool IsPathSnapshotDatadir(const fs::path& datadir_path)
+{
+    return fs::PathToString(datadir_path.filename()).length() == SNAPSHOT_NAME_LEN &&
+        fs::PathToString(datadir_path.filename()).substr(0,11) == "chainstate_";
+}
+
+static uint256 PathToSnapshotHash(const fs::path& datadir_path)
+{
+    assert(IsPathSnapshotDatadir(datadir_path));
+    return uint256S(fs::PathToString(datadir_path.filename()).substr(11));
+}
+
+static std::optional<fs::path> FindSnapshotChainstateDatadir()
+{
+    fs::path found;
+
+    for (fs::directory_iterator it(gArgs.GetDataDirNet()); it != fs::directory_iterator(); it++) {
+        if (fs::is_directory(*it) && !fs::is_empty(*it) && IsPathSnapshotDatadir(it->path()))
+        {
+            if (found.empty()) {
+                found = it->path();
+                LogPrintf("[snapshot] found snapshot datadir %s\n", fs::PathToString(found));
+            } else {
+                LogPrintf("[snapshot] WARNING - detected multiple snapshot " /* Continued */
+                    "datadirs (%s). This is not expected.\n", fs::PathToString(it->path()));
```

Comment:
> I think the best thing for bitcoin to do at startup would be to list all the chainstate* directories and use whichever one has the most work at its tip as the active chainstate.

I agree this would be a nice feature, but the process of unpacking snapshots and loading them to the point of being able to evaluate "most work" entails some big changes, so I'll defer this one for later. 

I do like your suggestion of calling it `<datadir>/chainstate_snapshot` and will incorporate that.

Reply:
I've changed the code to write the snapshot chainstate leveldb to `chainstate_snapshot`. A separate file within that directory called `base_blockhash` contains a serialized uint256 of what used to be in the directory name.